doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+
+The *rehydration time* dimension is effectively a classical batch database cost model: given cardinalities on the input relations, predict how much work a given plan has to do.
+
+The *latency per update* dimension is unique to our streaming setting. Different query plans will lead to different costs when different sources produce new data. If a calculus metaphor is helpful, you can think of the cost model here producing a partial derivative with respect to each input; alternatively, the cost model’s aggregate *latency per update* cost is the [Jacobian](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant) of that query.
+
+The *storage* dimension is meant to identify particularly expensive new arrangements: for example, a count query grouping by US state code (just 50, no big deal) vs. grouping by user ID (as big as your user table).
+
+The *scalability* dimension is meant to characterize how much the work for the given query can be spread across multiple nodes.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+This hasn't been implemented yet.
+
+# Rollout
+[rollout]: #rollout
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+Testing can be split into two categories:
+
+- Validation of the cost model in terms of a corpus of interesting/worthwhile queries.
+- Correctness of the cost model in terms of the costs it assigns to particular small queries.
+
+High-level validation requires running queries on meaningful amounts of data; low-level correctness checking might be able to run via `datadriven::walk` without using any data at all, just checking that certain queries are always given better costs than other queries.
+
+## Lifecycle
+[lifecycle]: #lifecycle
+
+This feature will only be visible as `EXPLAIN WITH(cost)`, with no real contract with users about the meaning of costs (at least for now).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A bad cost model may be worse than no cost model.
+
+# Conclusion and alternatives
+[conclusion-and-alternatives]: #conclusion-and-alternatives
+
+## Alternatives
+
+Cost models all look more or less the same, though the precise input features and outputs vary. These four factors and their input features are what we came up with as sensible things to go for at first.
+
+An alternative implementation approach would be to build a completely empirical cost model. That is, rather than writing a symbolic cost model *a priori*, we could curate a corpus of candidate queries, observe their performance on a variety of inputs, and then try to fit a cost model based on those results. Such an empirical cost model will almost surely be overfit and not particularly robust; it would also be hard to maintain.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Which loads should we test on? No cost model is perfect, but more testing with better/more realistic loads will give us a 
```

Comment:
Something is missing here!

Reply:
Apparently I took 'unresolved' a little too seriously here!